datamol/align.py

```diff
@@ -0,0 +1,252 @@
+    # so if the template is a subgraph of the molecule, no need to perform any MCS
+    # Another reason for this, is to avoid inconsistency in alignment between molecules that are
+    # supergraph of the template vs molecules that are subgraph of the template.
+    pattern = _template
+    if not _mol.HasSubstructMatch(_template):
+
+        pattern = None
+        mcs_smarts = dm.find_mcs([_mol, _template])
+
+        if mcs_smarts is not None:
+            pattern = dm.from_smarts(mcs_smarts)
+
+    if pattern is not None:
+
+        if auto_select_coord_gen:
+            rdDepictor.SetPreferCoordGen(use_depiction)
+
+        # we would need to compute 2d coordinates for the molecules if it doesn't have any
+        if _mol.GetNumConformers() == 0:
+            _mol = compute_2d_coords(_mol)
+
+        if use_depiction:
+            rdDepictor.GenerateDepictionMatching2DStructure(
+                _mol,
+                reference=_template,
+                refPatt=pattern,
+                acceptFailure=True,
+                allowRGroups=True,
+            )
+        else:
+            query_match = _mol.GetSubstructMatch(pattern)
+            template_match = _template.GetSubstructMatch(pattern)
+            rdMolAlign.AlignMol(_mol, _template, atomMap=list(zip(query_match, template_match)))
+
+    return _mol
+
+
+def auto_align_many(
+    mols: Union[Sequence[Mol], pd.Series],
+    partition_method: str = "anon-scaffold",
+    copy: bool = True,
+    cluster_cutoff: float = 0.7,
+    **kwargs,
+):
+    """Partition a list of molecules into clusters sharing common scaffold of common core,
+    then align the molecules to that common core. This function will compute the list of
+    smiles/smarts representative of each cluster first.
+
+    The returned molecules will have two properties associated to them:
+
+    - `dm.auto_align_many.cluster_id`: the cluster id of the molecule.
+    - `dm.auto_align_many.core`: the smiles/smarts of the core of the cluster.
+
+    Args:
+        mols: A list of molecules to auto align.
+        partition_method: Partition method to use.
+            One of ['scaffold', 'strip-scaffold', 'anongraph-scaffold', 'cluster' and 'anon-scaffold'].
+            Cautious as the method 'cluster' is very sensitive to the cutoff.
+        copy: Whether to copy the molecules before aligning them.
+        cluster_cutoff: Optional cluster cutoff.
+        kwargs: Additional arguments to pass to clustering method
+    """
+
+    if copy:
+        mols = [dm.copy_mol(mol) for mol in mols]
+
+    mol_groups = ddict(list)  # map scaffold index to list of unique molecules
+    scaffold_mols = {}
+
+    if partition_method.endswith("scaffold"):
+
+        scaffolds = [dm.to_scaffold_murcko(m) for m in mols]
+        scaffolds_ids = [dm.to_smiles(x) for x in scaffolds]
+
+        if partition_method.startswith("strip-"):
+            # with brics scaffold we compute brics bonds, then re-infer and select the biggest
```

Comment:
comment is outdated. The brics approach did not work well, so this just cut Rgroup according to some very opinionated smarts.

Reply:
will remove